### PR TITLE
Pytilt

### PR DESCRIPTION
**Pull request by jasperrxzhu**

### CODE REVIEW

File: python/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+set(CMAKE_CXX_EXTENSIONS OFF)
+
+find_package(LLVM 15 REQUIRED CONFIG)
+add_definitions(${LLVM_DEFINITIONS})
+
+add_subdirectory(pybind11)
+
+add_library(tilt STATIC IMPORTED)
+set_target_properties(tilt PROPERTIES
+    IMPORTED_LOCATION ${TILT_DIRECTORY}/build/src/libtilt.so)
```
Comment (anandj91) [inline python/CMakeLists.txt:18]: Don't link libraries like this. You can add the library in cmake. Check how google test is imported

File: python/eg.py [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+)
+
+pt.print_loopIR(sel_op, "select_loopIR.txt")
+
+
+### Example 2 ###
+### Print out the Loop IR for a Sum operator into sum_loopIR.txt ###
+def acc( s : pt.expr, st : pt.expr, et : pt.expr, d : pt.expr ) -> pt.expr :
+    return pt.add(s, d)
+acc_lambda = lambda s, st, et, d : pt.add(s, d)
```
Comment (anandj91) [inline python/eg.py:25]: Does pybind have conversion from lambdas to std::function?

File: python/setup.py [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+    def build_cmake_ext(self, ext):
+
+        if not os.path.exists(self.build_temp):
+            os.makedirs(self.build_temp)
+
+        extdir = os.path.abspath(os.path.dirname(self.get_ext_fullpath(ext.name)))
+
+        # manually change LLVM_DIR and TiLT_DIR for the time-being
+        LLVM_DIR = "/home/jasperzhu/llvm-project-15.0.7/build/lib/cmake/llvm"
+        TILT_DIR = "/home/jasperzhu/tilt"
```
Comment (anandj91) [inline python/setup.py:34]: lets not hardcode directory locations like this. make it parameterized. 
I don't think you need TILT_DIR location. You can get this from CMAKE_CURRENT_SOURCE_DIR env variable

**Review by anandj91 — CHANGES_REQUESTED** [review]

File: python/setup.py (hunk 3 above)
Comment (anandj91) [reply python/setup.py:34]: Also `setup.py` should be at the root directory. Keep only binding related source files inside `python/` directory.

File: .gitignore [hunk 4]
```diff
@@ -1,2 +1,3 @@
 .vscode
 build
+python/cleanup.sh
```
[the remark this answers is not in the document]
Comment (jasperrxzhu) [reply .gitignore]: Just a script I had to cleanup python virtual environments. Removed.

File: .gitmodules [hunk 5]
```diff
@@ -1,3 +1,6 @@
 [submodule "third_party/googletest"]
 	path = third_party/googletest
-	url = https://github.com/google/googletest.git
\ No newline at end of file
+	url = https://github.com/google/googletest.git
+[submodule "python/pybind11"]
+	path = python/pybind11
+	url = git@github.com:pybind/pybind11.git
```
[the remark this answers is not in the document]
Comment (jasperrxzhu) [reply .gitmodules]: Moved.

File: python/CMakeLists.txt (hunk 1 above)
Comment (jasperrxzhu) [reply python/CMakeLists.txt]: Changed how tilt was linked.

File: python/setup.py (hunk 3 above)
Comment (jasperrxzhu) [reply python/setup.py]: Moved setup.py to the root directory. LLVM_DIR is now specified using an environment variable.

File: python/CMakeLists.txt [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+set(CMAKE_CXX_COMPILER clang++)
+
+project(pytilt)
+
+set(CMAKE_CXX_STANDARD 17)
+set(CMAKE_CXX_STANDARD_REQUIRED ON)
+set(CMAKE_CXX_EXTENSIONS OFF)
+
+find_package(LLVM 15 REQUIRED CONFIG)
+add_definitions(${LLVM_DEFINITIONS})
```
[the remark this answers is not in the document]
Comment (jasperrxzhu) [reply python/CMakeLists.txt]: Removed.

File: python/CMakeLists.txt [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+
+add_subdirectory(pybind11)
+
+add_library(tilt STATIC IMPORTED)
+set_target_properties(tilt PROPERTIES
+    IMPORTED_LOCATION ${TILT_DIRECTORY}/build/src/libtilt.so)
+
+pybind11_add_module(pytilt main.cpp)
+target_include_directories(pytilt PUBLIC ${LLVM_INCLUDE_DIRS} ${TILT_DIRECTORY}/include)
```
[the remark this answers is not in the document]
Comment (jasperrxzhu) [reply python/CMakeLists.txt]: Removed.

File: CMakeLists.txt [hunk 8]
```diff
@@ -13,3 +13,9 @@ add_subdirectory(src)
 
 enable_testing()
 add_subdirectory(test)
+
+option(BUILD_PYTILT "Option description" OFF)
+if(BUILD_PYTILT)
```
Comment (anandj91) [inline CMakeLists.txt:18]: Is this necessary? can we just build pytilt every time we build the repo?

**Review by anandj91 — CHANGES_REQUESTED** [review]

**Pull request merged**